### PR TITLE
refactor(stream): return command execution after initial barrier collection

### PR DESCRIPTION
**Pull request by kwannoel**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Part of https://github.com/risingwavelabs/risingwave/issues/12771 and https://github.com/risingwavelabs/risingwave/issues/13803 and https://github.com/risingwavelabs/risingwave/issues/14471

For create mv, on initial barrier collection we want to update the catalog to the frontend.
As such, we provide a variant of `run_command` , `run_command_until_collect` which does so.

Then the caller of `run_command_until_collect` is responsible for:
1. Reporting the catalog to the fe. (Separate PR)
2. Polling the finish channel.

This PR does not change existing behaviour. Just does some refactoring work.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Comment by wenym1** [comment]
I am thinking that, shall we just deprecate the `finish_tx` in the notifier, because notifier is associated to a barrier, but the finish of create streaming job is beyond the scope of barrier. 

Then the `run_command` will always return only when barrier is collected.  If anyone is interested in the finish of create streaming job, we can provide a functionality in catalog manager to subscribe on the finish of create streaming job, where subscribers will get notified when the status of streaming job transit from creating to created.

**Comment by kwannoel** [comment]
> I am thinking that, shall we just deprecate the `finish_tx` in the notifier, because notifier is associated to a barrier, but the finish of create streaming job is beyond the scope of barrier.
> 
> Then the `run_command` will always return only when barrier is collected. If anyone is interested in the finish of create streaming job, we can provide a functionality in catalog manager to subscribe on the finish of create streaming job, where subscribers will get notified when the status of streaming job transit from creating to created.

Sounds good to me. Only create streaming job requires a special `subscribe`.

**Comment by kwannoel** [comment]
I found that currently in the barrier manager, it only marks the fragments as created. For table catalog, it relies on the stream manager to first receive the notification from the barrier manager, before it commits the catalogs.
So we can’t subscribe the catalog manager to know when a stream job is completed, since it doesn’t know that.
We also can’t subscribe the fragment manager, because it only marks fragments as created in the v1 metadata manager, not in v2. 

You may share more of your idea later. But I think we can merge this PR as is.

File: src/meta/src/barrier/schedule.rs [hunk 1]
```diff
@@ -315,6 +317,33 @@ impl BarrierScheduler {
         .map(|i| i[1])
     }
 
+    pub async fn run_command_until_collected(
```
Comment (StrikeW) [inline src/meta/src/barrier/schedule.rs:320]: It is not clear the difference with `run_command` below, which `Run a command and return when it's completely finished.`  May you add some comments to elaborate the meaning of `until_collected`?

Comment (kwannoel) [reply src/meta/src/barrier/schedule.rs:320]: Sorry, this PR is incomplete, and that code may just be deleted.

**Pull request closed without merging**